internal/venice-common/src/main/java/com/linkedin/venice/memory/HeapSizeEstimator.java

```diff
@@ -0,0 +1,188 @@
+   * 3. Polymorphism is ignored (i.e., we do not guess the size of any potential implementation of an abstraction)
+   *
+   * The intended usage of this function is to call it once per runtime per class of interest and to store the result in
+   * a static field. If the instances of this class (and any other class contained within it) contain no fields which
+   * are variable-sized object, no nullable objects, and no polymorphism, then the result of this function is
+   * effectively the heap size of each instance. If any of these conditions are not true, then a function should be
+   * implemented which uses this class overhead as a base and then makes adjustments, e.g.:
+   *
+   * 1. For any null pointer, we can subtract from the base a value equal to the output of this function for the type of
+   *    the field which is null.
+   * 2. For any variable-sized object, then the actual size has to be taken into account.
+   * 3. If polymorphism comes into play, then the specific sizes of actual instances need to take into account the
+   *    concrete classes they are made of.
+   *
+   * @throws {@link StackOverflowError} It should be noted that this function is recursive in nature and so any class
+   *         which contains a loop in its class graph will cause a stack overflow.
+   */
+  public static <T> int getClassOverhead(Class<T> c) {
+    if (c == null) {
+      throw new NullPointerException("The class param must not be null.");
+    }
+
+    if (c.isEnum()) {
+      /**
+       * Each enum value is stored statically, and so the per-instance overhead of enum fields is only the pointer size,
+       * already taken into account in the field loop below.
+       */
+      return 0;
+    }
+
+    if (c.isPrimitive()) {
+      return PRIMITIVE_SIZES.get(c);
+    }
+
+    int size = c.isArray() ? ARRAY_HEADER_SIZE : OBJECT_HEADER_SIZE;
+
+    size += overheadOfFields(c);
+
+    // TODO: Fix imprecision due to internal loss in parent class layouts
+    Class parentClass = c.getSuperclass();
+    while (parentClass != null) {
+      size += overheadOfFields(parentClass);
+      parentClass = parentClass.getSuperclass();
+    }
+
+    int finalSize = roundUpToNearestAlignment(size);
+
+    return finalSize;
+  }
+
+  /** Deal with alignment by rounding up to the nearest alignment boundary. */
+  private static int roundUpToNearestAlignment(int size) {
+    int partialAlignmentWindowUsage = size % ALIGNMENT_SIZE;
+    int waste = partialAlignmentWindowUsage == 0 ? 0 : ALIGNMENT_SIZE - partialAlignmentWindowUsage;
+    int finalSize = size + waste;
+    return finalSize;
+  }
+
+  private static int overheadOfFields(Class c) {
+    int size = 0;
+
+    /**
+     * {@link Class#getDeclaredFields()} returns the fields (of all visibility, from private to public and everything
+     * in between) of the class, but not of its parent class.
+     */
+    for (Field f: c.getDeclaredFields()) {
```

Comment:
Some of the fields are not visible. For example there are native references.

Reply:
Could you point me to a class (perhaps from the JDK) that has a native reference in it? I'd like to test that scenario and see what happens...

Reply:
Java.lang.Object :)

Reply:
Then I'm definitely not taking it into account 😅 ... where can I learn more about this?

Reply:
Alex Shipilev explained that in one of the videos (video is not in English...)